### DISCUSSION TITLE
객체지향 설계 5원칙에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#81] [OOP]
.

**Comment by J-I-H-O** [comment]
SOLID

**Comment by gyungchan-jo** [comment]
1. 단일 책임 원칙(Single Responsibility Principle, SRP):
한 클래스는 하나의 책임만 가져야 합니다.

2. 개방-폐쇄 원칙(Open-Closed Principle, OCP):
확장에는 열려 있어야 하고 수정에는 닫혀 있어야 합니다.
새로운 기능이 추가되거나 요구 사항이 변경되더라도 기존의 코드를 수정하지 않고 확장할 수 있어야 합니다.

3. 리스코프 치환 원칙(Liskov Substitution Principle, LSP):
서브타입은 언제나 자신의 기반 타입으로 교체할 수 있어야 합니다.
상속 관계에서 서브 클래스는 기반 클래스의 기능을 변경하지 않고 확장하여 사용해야 합니다.

4. 인터페이스 분리 원칙(Interface Segregation Principle, ISP):
클라이언트는 자신이 사용하지 않는 메서드에 의존하도록 강요받지 않아야 합니다.
인터페이스는 클라이언트에 필요한 기능만을 제공해야 합니다.

5. 의존 역전 원칙(Dependency Inversion Principle, DIP):
객체에서 어떤 Class를 참조해야 한다면, 그 Class를 직접 참조하는 것이 아니라 그 대상의 상위 요소(추상 클래스 or 인터페이스)로 참조하라는 원칙입니다.

**Comment by le2sky** [comment]
객체지향 설계 5원칙이라 불리는 SOLID는 다음 요소들의 앞글자를 따왔습니다. (아니 객체지향 설계 5원칙이 SOLID라고 불리는건가?!)

`단일 책임 원칙` : 클래스를 변경해야 하는 이유는 오직 하나뿐이어야 합니다.
`개방 폐쇄 원칙` : 확장에는 열려 있고, 변경에는 닫혀 있어야 합니다. 요구사항의 변경이나 추가가 발생하더라도, 기존 요소의 수정은 일어나지 말아야 하며, 기존 구성요소를 쉽게 확장해서 재사용할 수 있어야 합니다.
`리스코브 치환 원칙` : 서브 타입은 언제나 기반 타입으로 교체할 수 있어야 합니다. 즉, 서브 타입은 기반 타입이 약속한 규약을 지켜야 함을 의미합니다. 
`인터페이스 분리 원칙` :  아래에서 다시 설명
`의존성 역전 원칙` : 상위 수준의 모듈은 하위 수준의 모듈에 의존해서는 안되며, 모두 추상화에 의존해야 합니다. 또한, 추상화는 구체적인 사항에 의존해서는 안되며, 구체적인 사항은 추상화에 의존해야 합니다. 
모듈이란 크기와 상관없이 클래스, 패키지, 라이브러리와 같이 프로그램을 구성하는 임의의 요소를 의미합니다.

**Reply by le2sky** [reply]
Interface segregation Principle

- Don't depend on things that you don't need
- 사용하지 않지만 의존성을 가지고 있다면 해당 인터페이스가 변경되면 영향을 받습니다.
- 이로인해서, 독립적인 개발과 배포가 불가합니다.
- 사용하는 기능만 제공하도록 인터페이스를 분리하여, 한 기능에 대한 변경의 여파를 최소화할 수 있습니다.
- 클라이언트 입장에서 인터페이스를 분리하라는 원칙입니다.

![image](https://github.com/2024-woowacourse-study/level-interview/assets/39932141/d1031a8e-1f39-4a59-a9d0-1b67f7406860)

위 예제에서 Job 클래스는 많은 일을 하고 있어서, Job에 대한 많은 의존이 존재합니다.
해당 경우에 독립적인 배포와 개발이 어렵습니다. Cutter가 사용하는 Job의 기능이 변경된다면..
모든 Job의 클라이언트 코드가 영향을 받을 수 있습니다.

![image](https://github.com/2024-woowacourse-study/level-interview/assets/39932141/b376f05a-7d95-4fc1-988a-86d5e379ba63)

위와 같은 Fat 클래스(Job) 예시에서 다수의 인터페이스를 구현하도록 개선한다면, 
어떤 인터페이스에 변경이 발생하면, Job과 해당 인터페이스를 사용하는 서브 시스템만 영향을 받습니다.
